src/controller/CHIPDeviceController.cpp

```diff
@@ -805,7 +805,14 @@ CHIP_ERROR DeviceCommissioner::UnpairDevice(NodeId remoteDeviceId)
 {
     // TODO: Send unpairing message to the remote device.
 
-    FreeRendezvousSession();
+    VerifyOrReturnError(mState == State::Initialized, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mDeviceBeingPaired < kNumMaxActiveDevices, CHIP_ERROR_INCORRECT_STATE);
```

Comment:
This is still not correct, because we we'll return here if we're trying to unpair a device that is **already** paired AND there's no device **currently being** paired.

Reply:
How about change it to the following?

```
if (mDeviceBeingPaired < kNumMaxActiveDevices)
{
     Device * device = &mActiveDevices[mDeviceBeingPaired];
     if (device->GetDeviceId() == remoteDeviceId)
     {
         FreeRendezvousSession();
     }
}
```

Reply:
Lets start from here. Here is the logic:
UnpairDevice() should be able to unpair both a previously paired device or a device currently under pairing.

1. For previously paired device, mDeviceBeingPaired == kNumMaxActiveDevices, we don't need to call FreeRendezvousSession

2. For the device current under pairing, mDeviceBeingPaired < kNumMaxActiveDevices  and device->GetDeviceId() == remoteDeviceId
We need to call FreeRendezvousSession to prevent OnRendezvousComplete is called after RendezvousCleanup is called

Is this logic right?

Reply:
For previously paired device, mDeviceBeingPaired != remoteDeviceId. If it's < kNumMaxActiveDevices, we're pairing a different device. If it's == kNumMaxActiveDevices, then we're not pairing a different device.